File: WinForms/VB/HtmlMainDemo/Modules/RepositoryItems.vb
Imports System.Drawing
Imports DevExpress.Data.Utils.Security
Imports DevExpress.Utils.Html
Imports DevExpress.Utils.Svg
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Controls

Namespace DevExpress.HTML.Demos

    Public Partial Class RepositoryItems
        Inherits TutorialControl

        Private loginTemplate, signupTemplate As String

        Public Sub New()
            InitializeComponent()
            InitTemplates()
            InitCodeViewer()
            InitDataContext()
            EnsureSkinColors()
        End Sub

        Private Sub InitTemplates()
            loginTemplate = htmlContentControl.HtmlTemplate.Template
            signupTemplate = Properties.Resources.SignUpTemplate
        End Sub

        Private Sub InitCodeViewer()
            templateCodeViewer.Fill(htmlContentControl)
        End Sub

        Private Sub InitDataContext()
            Dim defaultPhoto = svgImageCollection1("default_photo")
            bindingSource.DataSource = New UserInfo(defaultPhoto)
            htmlContentControl.DataContext = bindingSource
        End Sub

        Private Sub OnHtmlElementMouseDown(ByVal sender As Object, ByVal e As DxHtmlElementMouseEventArgs)
            If Equals(e.ElementId, "signup") Then
                loginTemplate = htmlContentControl.HtmlTemplate.Template
                htmlContentControl.HtmlTemplate.Template = signupTemplate
                templateCodeViewer.UpdateHtmlTemplate()
            End If

            If Equals(e.ElementId, "login") Then
                signupTemplate = htmlContentControl.HtmlTemplate.Template
                htmlContentControl.HtmlTemplate.Template = loginTemplate
                templateCodeViewer.UpdateHtmlTemplate()
            End If

            If Equals(e.ElementId, "choose_Photo") Then
                htmlContentControl.ShowEditor("Photo")
                If htmlContentControl.ActiveEditor IsNot Nothing Then CType(htmlContentControl.ActiveEditor, PictureEdit).LoadImage()
            End If
        End Sub

        Private Sub OnPictureEditImageChanged(ByVal sender As Object, ByVal e As System.EventArgs)
            pictureEdit.OptionsMask.MaskType = If(htmlContentControl.ActiveEditor.EditValue IsNot Nothing, PictureEditMaskType.Circle, PictureEditMaskType.None)
        End Sub

        Protected Overrides Sub OnLookAndFeelChanged()
            EnsureSkinColors()
        End Sub

        Private Sub EnsureSkinColors()
            Dim checkedImage = checkEdit.ImageOptions.SvgImageChecked
            checkEdit.ImageOptions.SvgImageChecked = Nothing
            If checkedImage.DefaultStyle Is Nothing Then checkedImage.DefaultStyle = New SvgStyle()
            checkedImage.DefaultStyle.SetValue("fill", GetSkinColor("@HighlightAlternate"))
            checkEdit.ImageOptions.SvgImageChecked = checkedImage
        End Sub

        Private Function GetSkinColor(ByVal skinColorName As String) As String
            Return ColorTranslator.ToHtml(Utils.Colors.DXSkinColorHelper.GetDXSkinColor(Skins.ColorCreator.Create(skinColorName), LookAndFeel))
        End Function
    End Class

    Public Class UserInfo

        Private photoField As Byte()

        Private ReadOnly defaultPhoto As SvgImage

        Public Sub New(ByVal defaultPhoto As SvgImage)
            Me.defaultPhoto = defaultPhoto
        End Sub

        Public Property Email As String

        Private ReadOnly passwordData As SensitiveData = SensitiveData.CreateForCurrentUser()

        Public Property Password As String
            Get
                Return passwordData.Text
            End Get

            Set(ByVal value As String)
                passwordData.Text = value
            End Set
        End Property

        Private ReadOnly confirmData As SensitiveData = SensitiveData.CreateForCurrentUser()

        Public Property Confirm As String
            Get
                Return confirmData.Text
            End Get

            Set(ByVal value As String)
                confirmData.Text = value
            End Set
        End Property

        Public Property Agreement As Boolean

        Public ReadOnly Property IsDefaultPhoto As Boolean
            Get
                Return photoField Is Nothing
            End Get
        End Property

        Public ReadOnly Property HasPhoto As Boolean
            Get
                Return Not IsDefaultPhoto
            End Get
        End Property

        Public Property Photo As Object
            Get
                Return If(CObj(photoField), defaultPhoto)
            End Get

            Set(ByVal value As Object)
                If photoField Is value Then Return
                photoField = TryCast(value, Byte())
            End Set
        End Property
    End Class
End Namespace

File: WinForms/CS/HtmlMainDemo/Modules/RepositoryItems.cs
using System.Drawing;
using DevExpress.Data.Utils.Security;
using DevExpress.Utils.Html;
using DevExpress.Utils.Svg;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Controls;

namespace DevExpress.HTML.Demos {
    public partial class RepositoryItems : TutorialControl {
        string loginTemplate, signupTemplate;
        public RepositoryItems() {
            InitializeComponent();
            InitTemplates();
            InitCodeViewer();
            InitDataContext();
            EnsureSkinColors();
        }
        void InitTemplates() {
            loginTemplate = htmlContentControl.HtmlTemplate.Template;
            signupTemplate = Properties.Resources.SignUpTemplate;
        }
        void InitCodeViewer() {
            templateCodeViewer.Fill(htmlContentControl);
        }
        void InitDataContext() {
            var defaultPhoto = svgImageCollection1["default_photo"];
            bindingSource.DataSource = new UserInfo(defaultPhoto);
            htmlContentControl.DataContext = bindingSource;
        }
        void OnHtmlElementMouseDown(object sender, DxHtmlElementMouseEventArgs e) {
            if(e.ElementId == "signup") {
                loginTemplate = htmlContentControl.HtmlTemplate.Template;
                htmlContentControl.HtmlTemplate.Template = signupTemplate;
                templateCodeViewer.UpdateHtmlTemplate();
            }
            if(e.ElementId == "login") {
                signupTemplate = htmlContentControl.HtmlTemplate.Template;
                htmlContentControl.HtmlTemplate.Template = loginTemplate;
                templateCodeViewer.UpdateHtmlTemplate();
            }
            if(e.ElementId == "choose_Photo") {
                htmlContentControl.ShowEditor("Photo");
                if(htmlContentControl.ActiveEditor != null)
                    ((PictureEdit)htmlContentControl.ActiveEditor).LoadImage();
            }
        }
        void OnPictureEditImageChanged(object sender, System.EventArgs e) {
            pictureEdit.OptionsMask.MaskType = htmlContentControl.ActiveEditor.EditValue != null
                ? PictureEditMaskType.Circle : PictureEditMaskType.None;
        }
        protected override void OnLookAndFeelChanged() {
            EnsureSkinColors();
        }
        void EnsureSkinColors() {
            var checkedImage = checkEdit.ImageOptions.SvgImageChecked;
            checkEdit.ImageOptions.SvgImageChecked = null;
            if(checkedImage.DefaultStyle == null)
                checkedImage.DefaultStyle = new SvgStyle();
            checkedImage.DefaultStyle.SetValue("fill", GetSkinColor("@HighlightAlternate"));
            checkEdit.ImageOptions.SvgImageChecked = checkedImage;
        }
        string GetSkinColor(string skinColorName) {
            return ColorTranslator.ToHtml(DevExpress.Utils.Colors.DXSkinColorHelper.GetDXSkinColor(Skins.ColorCreator.Create(skinColorName), LookAndFeel));
        }
    }
    public class UserInfo {
        byte[] photo;
        readonly SvgImage defaultPhoto;
        public UserInfo(SvgImage defaultPhoto) {
            this.defaultPhoto = defaultPhoto;
        }
        public string Email { get; set; }
        readonly SensitiveData passwordData = SensitiveData.CreateForCurrentUser();
        public string Password {
            get { return passwordData.Text; }
            set { passwordData.Text = value; }
        }
        readonly SensitiveData confirmData = SensitiveData.CreateForCurrentUser();
        public string Confirm {
            get { return confirmData.Text; }
            set { confirmData.Text = value; }
        }
        public bool Agreement { get; set; }
        public bool IsDefaultPhoto => photo == null;
        public bool HasPhoto => !IsDefaultPhoto;
        public object Photo {
            get { return (object)photo ?? defaultPhoto; }
            set {
                if(photo == value)
                    return;
                photo = value as byte[];
            }
        }
    }
}
